### PR TITLE
front: converting train labels between osrd and nge

### LINKED ISSUE #8397: Converting labels between OSRD and NGE
### Description and goal

The goal of this feature is to share the labels of trains and nodes between OSRD and NGE

### Acceptance criteria

Trains:
- [x] If a label is added in OSRD, it is added in NGE
- [x] If a label is deleted in OSRD, it is deleted in NGE
- [x] If a label is added in NGE, it is added in OSRD
- [x] If a label is deleted in NGE, it is deleted in OSRD
- [x] If a label is updated in NGE, it is updated in OSRD

Macro nodes:
- [ ] If a label is added in OSRD, it is added in NGE
- [ ] If a label is deleted in OSRD, it is deleted in NGE
- [ ] If a label is added in NGE, it is added in OSRD
- [ ] If a label is deleted in NGE, it is deleted in OSRD
- [ ] If a label is updated in NGE, it is updated in OSRD

### PR DESCRIPTION
**Pull request by Uriel-Sautron**
first part of #8397

### CODE REVIEW

**Review by Caracol3 — APPROVED** [review]
lgtm (tested)

File: front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts]: To avoid doing another map at the end, maybe we can directly stock the ids in `formatedLabels` ?
I also feel like we have a lot of nesting loops. Maybe we can improve that ?

```suggestion
  const trainruns: Trainrun[] = trainSchedules.map((trainSchedule) => {
    let formatedLabels: (string | number)[] = [];
    if (trainSchedule.labels) {
      formatedLabels = trainSchedule.labels.map((label) => {
        const DTOLabel = DTOLabels.find((DTOlabel) => DTOlabel.label === label);
        if (DTOLabel) {
          return DTOLabel.id;
        }
        const newDTOLabel: NGELabel = {
          id: uuid(),
          label,
          labelGroupId: 1,
          labelRef: 'Trainrun',
        };
        DTOLabels.push(newDTOLabel);
        return newDTOLabel.id;
      });
    }
    return {
      id: trainSchedule.id,
      name: trainSchedule.train_name,
      categoryId: DEFAULT_TRAINRUN_CATEGORY.id,
      frequencyId: DEFAULT_TRAINRUN_FREQUENCY.id,
      trainrunTimeCategoryId: DEFAULT_TRAINRUN_TIME_CATEGORY.id,
      labelIds: formatedLabels,
    };
  });
```

**Review by SharglutDev — COMMENTED** [review]
Thank you for the PR, left just one comment

**Comment by louisgreiner** [comment]
Seems like there is something missing for the synchronization when we create a label from OSRD and then try to visualize it in NGE
image

File: front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts]: Done

File: front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts
Comment (emersion) [inline front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts]: Could we use an auto-incrementing ID (like other objects) instead of a string? That'd better match NGE's data type.

File: front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts
Comment (emersion) [inline front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:171]: There should always be a match here, no?

Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:171]: No, a train doesn't always have a label.

Comment (emersion) [reply front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:171]: But all label IDs in train runs must always have an entry in the dto labels array?

File: front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts]: Done

**Review by SharglutDev — APPROVED** [review]
Lgtm (ok for merge after @emersion approval :)

File: front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:171]: Yes, there will always be a match.
But i still have a TS error if I don't put "?"

File: front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts
Comment (louisgreiner) [inline front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts]:
```suggestion
          id: labelId++,
          label,
          labelGroupId: 1,
          labelRef: 'Trainrun',
        };
        DTOLabels.push(newDTOLabel);
```

File: front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts
Comment (emersion) [reply front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:171]: I'd suggest using `!` then, that way we'll get an error (which we can debug) in case this assumption ends up being broken, instead of silently ignoring the issue.

File: front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts
Comment (emersion) [reply front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts]: `++` is forbidden in our style guide (and I'm not a fan of using `++` as an expression instead of a statement, it's less clear whether the previous or next value ends up being used).

File: front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:171]: Ok, this solution also seems the best thank you for this catch.

Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts:171]: Done

File: front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts
Comment (SharglutDev) [reply front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts:307]: I think eslint doesn't link  de `++` :)

Comment (louisgreiner) [reply front/src/applications/operationalStudies/components/MacroEditor/osrdToNge.ts]: Good to know ! :)

**Comment by louisgreiner** [comment]
image

/!\ Labels are not displayed anymore on train schedules, even tho they still work (through filtering)

**Pull request merged**
